### PR TITLE
Fix endless reconciliation calls - 1.24.8

### PR DESCRIPTION
**Pull request by radufa**
**Description**

Changes proposed in this pull request:

- Fix endless reconciliation calls if the controller runs under "heavy load"

**Related issue(s)**

### CODE REVIEW

File: components/eventing-controller/reconciler/subscription-nats/reconciler.go [hunk 1]
```diff
@@ -164,102 +163,123 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+			eventingv1alpha1.ConditionReasonNATSSubscriptionActive, corev1.ConditionTrue, message)
 	}
 	for _, c := range sub.Status.Conditions {
 		var chosenCondition eventingv1alpha1.Condition
 		if c.Type == condition.Type {
-			// take given condition
-			chosenCondition = condition
-			conditionAdded = true
+			if !conditionContained {
+				if c.Status == condition.Status && c.Reason == condition.Reason && c.Message == condition.Message {
```
[the remark this answers is not in the document]
Comment (nachtmaar) [reply components/eventing-controller/reconciler/subscription-nats/reconciler.go:255]: I am also willing to contribute that 😂

**Review by nachtmaar — COMMENTED** [review]
Can you summarize the problems with the previous code and explain the fix which solves the endless reconciliation? Just some short notes on the PR description. 
You explained it to me in person and that is good, however we should document it for others as well. It might also help others who review the PR 😃
Except that I could verify that the `endless reconcilation` is fixed on a cluster. Tested by creating more than 35 subscriptions and restarting the controller. Without your changes, the restart of the controller did cause the endless reconcilation.

File: components/eventing-controller/reconciler/subscription-nats/reconciler.go [hunk 2]
```diff
@@ -164,102 +163,123 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
 	}
-	desiredSubscription.Status.Conditions = desiredConditions
-	desiredSubscription.Status.Ready = isNatsSubReady
 
-	if !reflect.DeepEqual(sub.Status, desiredSubscription.Status) {
-		err := r.Client.Status().Update(ctx, desiredSubscription, &client.UpdateOptions{})
+	if conditionsUpdated {
+		sub.Status.Conditions = desiredConditions
+		sub.Status.Ready = isNatsSubReady
+		err := r.Client.Status().Update(ctx, sub, &client.UpdateOptions{})
```
[the remark this answers is not in the document]
Comment (radufa) [reply components/eventing-controller/reconciler/subscription-nats/reconciler.go:282]: It's checked how you suggested in the caller code

File: components/eventing-controller/reconciler/subscription-nats/reconciler.go [hunk 3]
```diff
@@ -268,19 +288,19 @@ func (r Reconciler) syncSubscriptionStatus(ctx context.Context, sub *eventingv1a
 	return nil
 }
 
-func (r Reconciler) assertSinkValidity(sink string) error {
+func (r *Reconciler) assertSinkValidity(sink string) error {
```
[the remark this answers is not in the document]
Comment (radufa) [reply components/eventing-controller/reconciler/subscription-nats/reconciler.go:291]: It's the usual way to attach methods to the exposed reconciliation object, used in all places by us  too (in main too). It avoids an unnecessary conversion.

File: components/eventing-controller/pkg/handlers/nats.go [hunk 4]
```diff
@@ -125,7 +125,7 @@ func (n *Nats) SyncSubscription(sub *eventingv1alpha1.Subscription, cleaner even
 // DeleteSubscription deletes all NATS subscriptions corresponding to a Kyma subscription
 func (n *Nats) DeleteSubscription(subscription *eventingv1alpha1.Subscription) error {
 	for key, sub := range n.subscriptions {
-		if strings.HasPrefix(key, createKeyPrefix(subscription)) {
+		if createKeyPrefix(subscription) == createKymaSubscriptionNamespacedName(key, sub).String() {
```
Comment (nachtmaar) [inline components/eventing-controller/pkg/handlers/nats.go:128]: My proposal: put function into external function and add some table driven tests. I am willing to contribute them

**Comment by nachtmaar** [comment]
I found one more [call](https://github.com/radufa/kyma/blob/fix-endless-reconciliation-calls-1.24.8/components/eventing-controller/reconciler/subscription-nats/reconciler.go#L315) where the pattern
```
if k8serrors.IsConflict(err) {
    return ctrl.Result{Requeue: true}, nil
}
```
is not used yet. Let's have it consistent 😃

File: components/eventing-controller/reconciler/subscription-nats/reconciler.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nachtmaar) [reply components/eventing-controller/reconciler/subscription-nats/reconciler.go:255]: We agreed to perform this change on main instead.

File: components/eventing-controller/reconciler/subscription-nats/reconciler.go [hunk 5]
```diff
@@ -164,102 +163,123 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
-
-	_, err = r.Backend.SyncSubscription(desiredSubscription, r.eventTypeCleaner)
+	_, err = r.Backend.SyncSubscription(subscription, r.eventTypeCleaner)
 	if err != nil {
 		r.Log.Error(err, "failed to sync subscription")
-		if err := r.syncSubscriptionStatus(ctx, actualSubscription, false, err.Error()); err != nil {
-			return ctrl.Result{}, err
+		if err := r.syncSubscriptionStatus(ctx, subscription, false, err.Error()); err != nil {
+			if k8serrors.IsConflict(err) {
+				return ctrl.Result{Requeue: true}, nil
```
[the remark this answers is not in the document]
Comment (nachtmaar) [reply components/eventing-controller/reconciler/subscription-nats/reconciler.go:221]: We will add it to main instead.

**Review by nachtmaar — DISMISSED** [review]

**Review by nachtmaar — APPROVED** [review]
